addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRDevice.java

```diff
@@ -0,0 +1,481 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.dsmr.internal.device;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+
+import org.eclipse.smarthome.core.common.ThreadPoolManager;
+import org.openhab.binding.dsmr.DSMRBindingConstants;
+import org.openhab.binding.dsmr.internal.device.DSMRDeviceConstants.DSMRDeviceEvent;
+import org.openhab.binding.dsmr.internal.device.DSMRDeviceConstants.DSMRPortEvent;
+import org.openhab.binding.dsmr.internal.device.DSMRDeviceConstants.DeviceState;
+import org.openhab.binding.dsmr.internal.device.cosem.CosemObject;
+import org.openhab.binding.dsmr.internal.discovery.DSMRMeterDetector;
+import org.openhab.binding.dsmr.internal.discovery.DSMRMeterDiscoveryListener;
+import org.openhab.binding.dsmr.internal.meter.DSMRMeter;
+import org.openhab.binding.dsmr.internal.meter.DSMRMeterDescriptor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The DSMRDevice class represents the physical Thing within OpenHAB2 context
+ *
+ * The DSMRDevice will start in the state INITIALIZING where the serial port is opened.
+ *
+ * The device is waiting in the state STARTING till valid P1 telegrams are received. Then the device will enter the
+ * state ONLINE.
+ *
+ * During the STARTING state the class will switch baud rate automatically in case of communication problems.
+ *
+ * If there are communication problems or no P1 telegrams are received (anymore), the device will enter the state
+ * OFFLINE.
+ *
+ * If the OpenHAB2 system wants the binding to shutdown, the DSMRDevice enters the SHUTDOWN state and will close
+ * and release the OS resources.
+ *
+ * In case of configuration errors the DSMRDevice will enter the state CONFIGURATION_PROBLEM
+ *
+ * @author M. Volaart
+ * @since 2.1.0
+ */
+public class DSMRDevice implements DSMRPortEventListener {
+    // Logger
+    private final Logger logger = LoggerFactory.getLogger(DSMRDevice.class);
+
+    // DSMR Device configuration
+    private DSMRDeviceConfiguration deviceConfiguration = null;
+
+    // Device status class
+    private class DeviceStatus {
+        // State
+        private DeviceState deviceState;
+
+        // Timestamp when the device entered the deviceState
+        private long stateEnteredTS;
```

Comment:
Please make the field name as clear as the comment is, now I had to scroll back up here to understand the code.